wire/msgcfheaders.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2017 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package wire
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+const (
+	// MaxCFHeaderPayload is the maximum byte size of a committed
+	// filter header.
+	MaxCFHeaderPayload = chainhash.HashSize
+
+	// MaxCFHeadersPerMsg is the maximum number of committed filter headers
+	// that can be in a single cfheaders message.
+	MaxCFHeadersPerMsg = 2000
+)
+
+// MsgCFHeaders implements the Message interface and represents a cfheaders
+// message.  It is used to deliver committed filter header information in
+// response to a getcfheaders message (MsgGetCFHeaders). The maximum number of
+// committed filter headers per message is currently 2000. See MsgGetCFHeaders
+// for details on requesting the headers.
+type MsgCFHeaders struct {
+	StopHash     chainhash.Hash
+	FilterType   FilterType
+	HeaderHashes []*chainhash.Hash
+}
+
+// AddCFHeader adds a new committed filter header to the message.
+func (msg *MsgCFHeaders) AddCFHeader(headerHash *chainhash.Hash) error {
+	if len(msg.HeaderHashes)+1 > MaxCFHeadersPerMsg {
+		str := fmt.Sprintf("too many block headers in message [max %v]",
+			MaxBlockHeadersPerMsg)
+		return messageError("MsgCFHeaders.AddCFHeader", str)
+	}
+
+	msg.HeaderHashes = append(msg.HeaderHashes, headerHash)
+	return nil
+}
+
+// BtcDecode decodes r using the wire protocol encoding into the receiver.
+// This is part of the Message interface implementation.
+func (msg *MsgCFHeaders) BtcDecode(r io.Reader, pver uint32) error {
+	// Read stop hash
```

Comment:
All of these `BtcDecode` and `BtcEncode` functions for the new messages should be checking the protocol version and returning an error if it's before the minimum supported one.

e.g.

```
if pver < NodeCFVersion {
	str := fmt.Sprintf("getcftypes message invalid for protocol "+
		"version %d", pver)
	return messageError("MsgCFHeaders.BtcDecode", str)
}

...
```

Then, in the tests for all of these new messages, there should be a cross protocol test function (e.g. `TestCFHeadersCrossProtocol` for this case) which ensures attempting to decode a message encoded with the new protocol version fails as expected when attempting to decode it with an older protocol version.

See `msgfilteradd.go` and `msgfilteradd_test.go` in the upstream btcd code for examples.

Reply:
There appears to be variation in the meaning of the cross protocol tests.  In dcrd's TestSendHeadersCrossProtocol test, the two protocols used both support the message being tested and serialization/deserialization is tested to not error on either.